### PR TITLE
Added Papers, a video and Code

### PR DESCRIPTION
**Pull request by jejjohnson**
Added a lot of papers, a video and some code added!

**Questions**

* Not sure how the dates for the code are organized. So I didn't add it.
* Maybe we can separate repositories with actual packages? 
* Might be nice to highlight the back-end package for the code (e.g. PyTorch, JAX, TensorFlow, etc)
* The date convention, are we doing american (month-day-year) or the world (day-month-year)? A few follow the both conventions. 
* Maybe we could separate the papers into **algorithms** versus **applications**. There aren't that many applications from what I've seen. So perhaps this could help promote them! :)

### CODE REVIEW

**Comment by janosh** [comment]
Awesome! Thanks a lot for this great contribution!

Regarding your questions:

> Not sure how the dates for the code are organized. So I didn't add it.

Good point. The dates are quite a mess. 😞 

> Maybe we can separate repositories with actual packages?

That's a good idea! 👍 

> Might be nice to highlight the back-end package for the code (e.g. PyTorch, JAX, TensorFlow, etc)

Another great suggestion. Let's start new subheadings for each library under the `## Code` heading as well as under a new heading `## Packages`.

> The date convention, are we doing american (month-day-year) or the world (day-month-year)? A few follow the both conventions.

Let's go with the international format of day-month-year.

> Maybe we could separate the papers into algorithms versus applications. There aren't that many applications from what I've seen. So perhaps this could help promote them! :)

Hmmm. What do you mean by applications? Most of the algorithm papers also go on to apply their method. Would those go in both? Or do you mean applied to some real-world problem?

Let me know which of your ideas you'd like to submit PRs for. Otherwise, I'll try and tackle them as time permits.

**Comment by janosh** [comment]
Just fixed the date formatting across all items. Went with the American format after all because having the dates first right after the numbered list was hard to read (e.g. `4. 5 Feb 2018` is worse than `4. Feb 5, 2018`).

**Comment by janosh** [comment]
Splitting `## Code` and the new `## Packages` section into subsections based on backend (PyTorch, TensorFlow, JAX for now) is done.

**Comment by jejjohnson** [comment]
I guess since you did most of the stuff I suggested, I'll try and see if I can find some applications. A few that come to mind are:
* [Super Resolution Flows](https://arxiv.org/abs/2006.14200)
* [Flows for simultaneous manifold learning and density estimation](https://arxiv.org/abs/2003.13913)
* [Faster Uncertainty Quantification for Inverse Problems with Conditional Normalizing Flows](https://arxiv.org/abs/2007.07985)
* [Analyzing Inverse Problems with Invertible Neural Networks](https://arxiv.org/abs/1808.04730)

So I wouldn't say that these are developing new NF methods persay but instead finding creative ways to use them. But I could be mistaken. Perhaps for now, just add them and later we can make more sub-categories. **OR** we can just put a theory, model or application tag.

**Comment by janosh** [comment]
Those would be great additions! Happy to accept a PR.

File: readme.md [hunk 1]
```diff
@@ -100,6 +130,38 @@ A list of awesome resources for understanding and applying normalizing flows (NF
...
+   > A great repo with some basic implementations of normalizing flows from scratch.
+
+2. [`nflows`](https://github.com/bayesiains/nflows) by [bayesiains](https://homepages.inf.ed.ac.uk/imurray2/group/)
+    > A suite of most of the SOTA methods using PyTorch. From an ML group in Edinburgh. They created the current SOTA spline flows. Almost as complete as you'll find from a single repo.
+
+3. [`Normalizing Flows - Introduction (Part 1)`](https://pyro.ai/examples/normalizing_flows_i.html) by [pyro.ai](http://pyro.ai/) 
+  
+    > A tutorial about how to use the `pyro-ppl` library (based on PyTorch) to use Normalizing flows. They have some of the SOTA methods including NSF and MAF.
+
+4. [`destructive-deep-learning`](https://github.com/davidinouye/destructive-deep-learning) by [David Inouye](https://www.davidinouye.com/)
```
Comment (janosh) [inline readme.md:144]: @jejjohnson Just added the backends to each of the codebases listed under `## Others`. After briefly looking at this one, I don't understand how it relates to NFs. Could you explain what you mean by

> An entire suite of iterative methods to normalizing flows.

Comment (jejjohnson) [reply readme.md:144]: @janosh Yeah sorry, so this is the author's code base for the Deep Density Destructors by Inouye & Ravikumar (2018) paper. It's essentially the same as normalising flows but you converge iteratively instead of fully parameterised. So you're still using bijective transforms, still computing the Jacobian, still checking for the likelihood, and you can still sample and get probability density estimates. The only difference is that you:

1.  compute one layer or block layer (e.g. Marginal Gaussianization + PCA rotation)
2. check for convergence (e.g log-likelihood using the change-of-variables formula)
3. repeat steps 1-2 until convergence.

To be fair, after the new updates, I would put this under author's code instead of a dedicated section as it's a bit misleading.

---
**Table 1** in the [paper](http://proceedings.mlr.press/v80/inouye18a/inouye18a.pdf) has a nice interesting comparison with the traditional NFs.

Comment (janosh) [reply readme.md:144]: Thanks for the clarification. I'll just add this to the readme.
